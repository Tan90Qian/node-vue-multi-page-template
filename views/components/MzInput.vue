<template>
  <div class="mz-input">
    <div class="mz-input-prefix"><slot name="prefix"></slot></div>
    <input
      ref="input"
      :type="type"
      :placeholder="placeholder"
      :value="value"
      @input="$emit('input', $event.target.value);"
      @focus="$emit('onFocus');"
      @blur="$emit('onBlur');"
      @keyup.enter="$emit('onEnter');"
    />
    <div class="mz-input-suffix"><slot name="suffix"></slot></div>
  </div>
</template>

<script>
export default {
  name: 'MzInput',
  props: {
    type: {
      type: String,
      default: 'text'
    },
    value: {
      type: String
    },
    placeholder: {
      type: String,
      default: '请输入'
    }
  }
};
</script>

<style lang="scss">
.mz-input {
  position: relative;
  input {
    border: 0;
    outline: none;
  }
  .mz-input-prefix,
  .mz-input-suffix {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
  }
  .mz-input-prefix {
    left: 0;
  }
  .mz-input-suffix {
    right: 0;
  }
}
</style>
