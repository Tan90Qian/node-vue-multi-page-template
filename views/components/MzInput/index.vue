<template>
  <span class="mz-input">
    <span v-if="$slots.prefix" class="mz-input-prefix"> <slot name="prefix"></slot> </span>
    <input
      ref="input"
      v-bind="$attrs"
      :type="type"
      :placeholder="placeholder"
      :value="stateValue"
      v-on="listeners"
    />
    <span v-if="$slots.suffix" class="mz-input-suffix"> <slot name="suffix"></slot> </span>
  </span>
</template>

<script>
import { hasProp } from '../_util/prop-util.js';
import { isIE, isIE9 } from '../_util/env.js';

function fixControlledValue(value) {
  if (typeof value === 'undefined' || value === null) {
    return '';
  }
  return value;
}

export default {
  name: 'MzInput',
  inheritAttrs: false,
  props: {
    type: {
      type: String,
      default: 'text'
    },
    value: String,
    defaultValue: String,
    autoFocus: Boolean,
    placeholder: {
      type: String,
      default: '请输入'
    }
  },
  model: {
    prop: 'value',
    event: 'change.value'
  },
  data() {
    const { value, defaultValue } = this.$props;
    return {
      stateValue: fixControlledValue(!hasProp(this, 'value') ? defaultValue : value)
    };
  },
  mounted() {
    this.$nextTick(() => {
      if (this.autoFocus) {
        this.$refs.input.focus();
      }
    });
  },
  watch: {
    value(val) {
      this.stateValue = fixControlledValue(val);
    }
  },
  methods: {
    handleKeyDown(e) {
      if (e.keyCode === 13) {
        this.$emit('onEnter', e);
      }
      this.$emit('keydown', e);
    },
    handleChange(e) {
      if (isIE && !isIE9 && this.stateValue === e.target.value) {
        return;
      }
      if (!hasProp(this, 'value')) {
        this.stateValue = e.target.value;
      } else {
        // this.$forceUpdate();
        if (!e.target.composing) {
          this.$emit('change.value', e.target.value);
        }
        this.$emit('change', e);
        this.$emit('input', e);
      }
    },
    handleFocus() {
      this.$emit('onFocus');
    },
    handleBlur() {
      this.$emit('onBlur');
    }
  },
  computed: {
    listeners() {
      const { $listeners, handleChange, handleKeyDown, handleFocus, handleBlur } = this;
      return {
        ...$listeners,
        focus: handleFocus,
        blur: handleBlur,
        input: handleChange,
        keydown: handleKeyDown
      };
    }
  }
};
</script>

<style lang="scss">
.mz-input {
  position: relative;
  input {
    border: 0;
    outline: none;
  }
  .mz-input-prefix,
  .mz-input-suffix {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
  }
  .mz-input-prefix {
    left: 0;
  }
  .mz-input-suffix {
    right: 0;
  }
}
</style>
