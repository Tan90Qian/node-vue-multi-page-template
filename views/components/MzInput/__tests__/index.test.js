import { mount } from '@vue/test-utils';
import Vue from 'vue';

import MzInput from '../index';

describe('MzInput tests', () => {
  /* input tests */
  it('render input', () => {
    const wrapper = mount(MzInput);

    expect(wrapper.contains('input')).toBe(true);
  });

  it('should render password input when type was password', () => {
    const wrapper = mount(MzInput, {
      propsData: {
        type: 'password'
      }
    });
    const input = wrapper.find({ ref: 'input' });

    expect(input.attributes('type')).toBe('password');
  });

  it('should set placeholder on input', () => {
    const wrapper = mount(MzInput, {
      propsData: {
        placeholder: 'please'
      }
    });
    const input = wrapper.find({ ref: 'input' });

    expect(input.attributes('placeholder')).toBe('please');
  });

  it('input should has data-test attr', () => {
    const wrapper = mount(MzInput, {
      attrs: {
        'data-test': '123'
      }
    });
    const input = wrapper.find({ ref: 'input' });

    expect(input.attributes('data-test')).toBe('123');
  });

  it('should emit onFocus event when input focus', () => {
    const wrapper = mount(MzInput);
    const input = wrapper.find({ ref: 'input' });
    input.trigger('focus');

    expect(wrapper.emitted().onFocus).toBeTruthy();
  });

  it('should emit onBlur event when input blur', () => {
    const wrapper = mount(MzInput);
    const input = wrapper.find({ ref: 'input' });
    input.trigger('blur');

    expect(wrapper.emitted().onBlur).toBeTruthy();
  });

  it('should emit onFocus once mounted when autoFocus given', () => {
    const wrapper = mount(MzInput, {
      propsData: {
        autoFocus: true
      }
    });

    return Vue.nextTick().then(() => {
      expect(wrapper.emitted().onFocus).toBeTruthy();
    });
  });

  /* slot tests */
  it('should not render prefix div when prefix slot absent', () => {
    const wrapper = mount(MzInput);

    expect(wrapper.contains('.mz-input-prefix')).toBe(false);
  });

  it('should render prefix div when prefix slot given', () => {
    const wrapper = mount(MzInput, {
      slots: {
        prefix: ['<i>123</i>']
      }
    });

    expect(wrapper.html()).toContain('<span class="mz-input-prefix"><i>123</i></span>');
  });

  it('should not render suffix div when suffix slot absent', () => {
    const wrapper = mount(MzInput);

    expect(wrapper.contains('.mz-input-suffix')).toBe(false);
  });

  it('should render suffix div when suffix slot given', () => {
    const wrapper = mount(MzInput, {
      slots: {
        suffix: ['<i>123</i>']
      }
    });

    expect(wrapper.html()).toContain('<span class="mz-input-suffix"><i>123</i></span>');
  });
});
