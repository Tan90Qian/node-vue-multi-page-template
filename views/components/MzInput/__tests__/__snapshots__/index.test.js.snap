// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MzInput tests renders correctly 1`] = `
<span
  class="mz-input"
>
  <!---->
   
  <input
    placeholder="请输入"
    type="text"
  />
   
  <!---->
</span>
`;
